spot-client/src/common/webrtc/PeerConnection.js

```diff
@@ -0,0 +1,231 @@
+import { Emitter } from 'common/emitter';
+import { logger } from 'common/logger';
+
+
+export const DEFAULT_STUN_SERVERS = [
+    { urls: 'stun:stun.l.google.com:19302' },
+    { urls: 'stun:stun1.l.google.com:19302' },
+    { urls: 'stun:stun2.l.google.com:19302' }
+];
+
+/**
+ * A wrapper around {@code RTCPeerConnection}. Currently only deals with a single data channel and does not handle
+ * any media streams.
+ */
+export default class PeerConnection extends Emitter {
+    /**
+     * Event sent when a new message is received over the data channel.
+     * @type {string}
+     */
+    static DATA_CHANNEL_MSG_RECEIVED = 'PC_DATA_CHANNEL_MSG_RECEIVED';
+
+    /**
+     * The {@code PeerConnection} class tries to manage a single data channel and this event is emitted whenever this
+     * data channel's status changes.
+     * @type {string}
+     */
+    static DATA_CHANNEL_STATUS_CHANGED = 'PC_DATA_CHANNEL_STATUS_CHANGED';
+
+    /**
+     * Event emitted when new ICE candidate has been discovered. It needs to be transferred over and set on the remote
+     * {@code PeerConnection}. See {@link PeerConnection.addRemoteIceCandidate}.
+     * @type {string}
+     */
+    static ON_ICE_CANDIDATE = 'PC_ON_ICE_CANDIDATE';
+
+    /**
+     * Creates new instance of {@link PeerConnection}.
+     *
+     * @param {string} remoteAddress - A string which identifies the remote peer.
+     * @param {RTCIceServer[]} [iceServers] - An array with ICE STUN/TURN servers description as defined by WebRTC. See
+     * {@link https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer/urls} for more details about the format.
+     */
+    constructor(remoteAddress, iceServers) {
+        super();
+        this.remoteAddress = remoteAddress;
+        this._iceServers = iceServers || DEFAULT_STUN_SERVERS;
+        this._peerConnection = null;
+        this._dataChannel = null;
+        this._currentStatus = false;
+    }
+
+    /**
+     * Call this method to add remote ICE candidate to this connection.
+     *
+     * @param {RTCIceCandidateInit} candidate - Remote ICE candidate to be added to this connection.
+     * @returns {Promise<void>}
+     */
+    addRemoteIceCandidate(candidate) {
+        return this._peerConnection.addIceCandidate(candidate);
+    }
+
+    /**
+     * Creates an offer and returns a structure that needs to be transferred over to the remote peer and set using
+     * {@link setOffer}.
+     *
+     * @returns {Promise<RTCSessionDescriptionInit>}
+     */
+    createOffer() {
+        this._initPeerConnection();
```

Comment:
What do you think about creating the peer connection in the constructor? That way more of the initialization happens while other instance variables are getting set. That would also prevent the case of calling createOffer while a peer connection already exists.

Reply:
I will look into that